### PR TITLE
feat: add pagination nav component

### PR DESCRIPTION
**Pull request by janhassel**
React implementation of PaginationNav component. https://the-carbon-components.netlify.app/?nav=pagination-nav

@joshblack We talked about this one a while ago, especially the dynamic behaviour of the overflow. I opened this as a draft PR so we can get a discussion started on what you guys need from us for this contribution.

---

This PR comes with two UX teaks in comparison to the current vanilla component.
1. It adds a "loop" prop (disabled by default) that allows users to cycle through all pages when they reach the end / start.
  - Screen Recording 2020-06-10 at 17 34 52
2. The behaviour of how the overflow is treated and presented to the user changes. Instead of overlaying the overflow with a page indicator, the component always keeps the current page visible and out of the overflow (see screen recordings attached)
  - Current behaviour: Screen Recording 2020-06-10 at 17 38 52
  - Proposed behaviour: Screen Recording 2020-06-10 at 17 40 32

### CODE REVIEW

**Comment by emyarod** [comment]
I've had this completed for almost 9 months now, [including the overflow behavior](https://github.com/carbon-design-system/carbon/compare/master...emyarod:2634-add-pagination-nav) but since it was deprioritized, all I have left is the test suite. if we're ready to bring this into the library I can just round out the test suite and open a PR

File: packages/react/src/components/PaginationNav/PaginationNav.js [hunk 1]
```diff
@@ -0,0 +1,363 @@
...
+    backHidden = backHidden <= totalItems - 3 ? backHidden : totalItems - 3;
+  }
+
+  return {
+    front: frontHidden,
+    back: backHidden,
+  };
+};
+
+export default class PaginationNav extends Component {
```
Comment (joshblack) [inline packages/react/src/components/PaginationNav/PaginationNav.js]: Quick style-nit, for net-new code we try and use function components with hooks 👀

File: packages/react/src/components/PaginationNav/PaginationNav.js [hunk 2]
```diff
@@ -0,0 +1,363 @@
...
+    if (previousIndex < 0) {
+      if (loop) {
+        this.jumpToItem(totalItems - 1);
+      }
+    } else {
+      this.jumpToItem(previousIndex);
+    }
+  }
+
+  renderDirectionButton(direction) {
```
Comment (joshblack) [inline packages/react/src/components/PaginationNav/PaginationNav.js]: Would it make sense for this to be a component that exists in the `PaginationNav` file?

File: packages/react/src/components/PaginationNav/PaginationNav.js [hunk 3]
```diff
@@ -0,0 +1,363 @@
...
+    const { page, cuts } = this.state;
+
+    const classNames = classnames(`${prefix}--pagination-nav`, className);
+
+    // if less than 5 items can be shown, the first overflow
+    // should appear at index 0, not 1 as usual
+    const startOffset = itemsShown < 5 && page > 1 ? 0 : 1;
+
+    return (
+      <nav className={classNames} {...other} aria-label="pagination" key={page}>
```
Comment (joshblack) [inline packages/react/src/components/PaginationNav/PaginationNav.js]: Quick question, what is `key` used for on the `nav` here 👀  Would this trigger reconciliation each time `page` is updated?

**Comment by joshblack** [comment]
@aagonzales gave it a look and really liked the changes! I think we could bring in the UX changes along with the code formatting stuff mentioned above (along with style guide stuff). Feel free to use any of the existing work too and we can definitely help ya out 😄 

If you have any questions about tests btw, just let me know!

File: packages/react/src/components/PaginationNav/PaginationNav.js (hunk 3 above)
Comment (janhassel) [reply packages/react/src/components/PaginationNav/PaginationNav.js]: That was indeed the intention, but I believe it was actually already obsolete at this point. It is definitely now with the refactor to the functional component.

**Comment by janhassel** [comment]
@joshblack Thanks for the feedback (and the addition of translateWithId in the styleguide). I refactored the component accordingly. Would you mind taking another look and if there are only minor things left, I can work on the test suite.

File: packages/react/src/components/PaginationNav/PaginationNav.js [hunk 4]
```diff
@@ -0,0 +1,341 @@
...
+}) {
+  if (count > 1) {
+    return (
+      <li className={`${prefix}--pagination-nav__list-item`}>
+        <div className={`${prefix}--pagination-nav__select`}>
+          {/* eslint-disable-next-line jsx-a11y/no-onchange */}
+          <select
+            className={`${prefix}--pagination-nav__page ${prefix}--pagination-nav__page--select`}
+            aria-label={`Select ${t('carbon.pagination-nav.item')} number`}
+            onChange={e => {
```
Comment (joshblack) [inline packages/react/src/components/PaginationNav/PaginationNav.js]: Should we be using `onBlur` here per the `jsx-a11y` rule?

**Review by joshblack — COMMENTED** [review]
Looking great! 🎉  Just had a couple of questions:

- What should the responsive behavior be like for this component? 
- Do we need to flag the current page with `aria-current="page"`?

For determining cuts, it seems like it works great at the midpoint but when hitting past that it seems to make what is clicked jump to the edge before the overflow:

demo

Would it make sense to have that midpoint behavior happen throughout the range of pages?

**Comment by janhassel** [comment]
@joshblack 
- I updated (and actually simplified) the overflow behaviour to feel more natural. Since I haven't worked with hooks very much I'd really appreciate your input on my implementation there since I suppressed the warnings from React about the dependency array and I'm not sure if there is another way for the functionality here.
  - Screen Recording 2020-06-19 at 17 34 14
- Regarding `aria-current="page"`: good point! Missed that, thanks!
- For the responsive behaviour our team is planning to always maximize it to width available. The exact calculation would probably depend on the product though. So passing in the maximum amount of pages to `props.itemsShown` would be my proposal. Btw, the component enforces a minimum of 4 shown items. Together with the two buttons that's 288px (6*48) width which fits perfectly in the smallest screen size supported (288px component + $spacing-05 left and $spacing-05 right = 320px)
  - Screen Recording 2020-06-19 at 17 48 57

File: packages/react/src/components/PaginationNav/PaginationNav.js (hunk 4 above)
Comment (janhassel) [reply packages/react/src/components/PaginationNav/PaginationNav.js]: The issue with `onBlur` is that it only triggers once the user clicks on another element or continues with <kbd>tab</kbd>

Screen Recording 2020-06-19 at 17 41 58

**Comment by joshblack** [comment]
Hey @janhassel! 👋  So glad that this got moved from draft -> ready for review 🎉 

I wanted to ask you if you had an open slot sometime this week for us to go over the PR on a call? Thought it might be helpful for all of us and hopefully speed up the process of getting things merged in 😄  This really would only be talking through the component, its intended behaviors, any relevant implementation details you think are worth sharing, etc. Totally understand if not! Just wanted to reach out and see.

**Comment by janhassel** [comment]
@joshblack Definitely! I sent you an invite for tomorrow morning, feel free to counter if you prefer a different time slot 🙂

**Review by aledavila — CHANGES_REQUESTED** [review]
This is awesome work. So far it looks good

Good to go: 
MacOS: Chrome, Firefox, and Safari. No DAP violations. Keyboard navigation works. 

Issues on VO:
Safari: 
When selecting a page it doesn't announce if I have selected it. If I am navigating via arrows it does not read which page I am highlighting.

**Review by dakahn — CHANGES_REQUESTED** [review]
Found a few things _testing with JAWS 2020 and NVDA on Firefox latest in Windows 10_:

- the active button has no focus state
- the current active page button is described as "unavailable" -- which makes sense technically, but sounds quite confusing
- when the user selects a page the button's label should be updated to 'active'
- the next page button when activated doesn't announce the changes (updated page from previous page)

**Comment by janhassel** [comment]
Thanks for the feedback!

@aledavila 
- Should be fixed now

@dakahn
- it's in now but it doesn't look too good due to the blue underline. Maybe @aagonzales can give some visual guidance here? 
  - image
- This is an interesting one as I couldn't provoke a similar behaviour on VO (the only machine with Windows and JAWS installed is in the office, so I currently don't have access to that one). Is this how JAWS usually describes a disabled button (similar to "dimmed" on VO) or what could cause that?
- I updated the label. Let me know if it is better now
- Should be fixed

I'm currently facing the problem that the live region "Page 5 of 10" is read twice and after some poking around it seems that the component is rendering twice when a button is clicked. However, I cannot see this in the React DevTools Profiler.
Do you guys have any idea what could be causing this (or is it maybe only an issue with my setup)?

**Comment by aagonzales** [comment]
Hi @janhassel ! I talked to Jeannie about the selected+focus state overlap and she says its fine. We agree it does look not the best but we can live with it since its a temporary state and follows the correct styles.

**Review by aledavila — APPROVED** [review]
Thanks for the updates! LGTM pending @dakahn changes

**Comment by dakahn** [comment]
We're reading button updates now with NVDA which is great. When the 'next page button' is activated you just here '5', '6' ,'7' etc maybe something more clear might be, "page 5, active", but this isn't a blocker.

Unavailable is what's read by JAWS and NVDA in place of dimmed -- it's not a blocker technically since it's correct.

~~JAWS 2020 still isn't reading the page updates when the 'next page button' is activated. Let me see if I can get this working for you since I can test on JAWS here on my machine~~

**Review by dakahn — APPROVED** [review]
We're actually good to go here. JAWS is reading aria-live updates just fine. (fun fact, JAWS will not read `aria-live` updates unless it is started and running before opening Firefox -- I'm talking to you, future me)

**Comment by janhassel** [comment]
@dakahn That's good to know actually, might save me a lot of time debugging once I have access to the JAWS machine again 😀

I added `aria-atomic` to the live region which should cause JAWS to read out the full status now ("Page 5 of 10") instead of just the number. I didn't notice a difference on Chrome or Safari with VO, hence I didn't include it earlier. Can you give it a try and see if it's as expected now?

**Pull request merged**